src/provider/form-factory-provider.tsx

```diff
@@ -0,0 +1,138 @@
+  location: OpenmrsResource;
+  provider: OpenmrsResource;
+  visit: OpenmrsResource;
+  children: React.ReactNode;
+  formSubmissionProps: {
+    isSubmitting: boolean;
+    setIsSubmitting: (isSubmitting: boolean) => void;
+    onSubmit: (data: any) => void;
+    onError: (error: any) => void;
+    handleClose: () => void;
+  };
+  setCurrentPage: (page: string) => void;
+  handleConfirmQuestionDeletion?: (question: Readonly<FormField>) => Promise<void>;
+}
+
+const FormFactoryProviderContext = createContext<FormFactoryProviderContextProps | undefined>(undefined);
+
+export const FormFactoryProvider: React.FC<FormFactoryProviderProps> = ({
+  patient,
+  sessionMode,
+  sessionDate,
+  formJson,
+  workspaceLayout,
+  location,
+  provider,
+  visit,
+  children,
+  formSubmissionProps,
+  setCurrentPage,
+  handleConfirmQuestionDeletion,
+}) => {
+  const rootForm = useRef<FormContextProps>();
+  const subForms = useRef<Record<string, FormContextProps>>({});
+  const layoutType = useLayoutType();
+  const { isSubmitting, setIsSubmitting, onSubmit, onError, handleClose } = formSubmissionProps;
+  const registerForm = (formId: string, context: FormContextProps, isSubForm: boolean) => {
+    if (isSubForm) {
+      subForms.current[formId] = context;
+    } else {
+      rootForm.current = context;
+    }
+  };
+  // TODO: Manage and load processors from the registry
+  const formProcessors = useRef<Record<string, FormProcessorConstructor>>({
+    EncounterFormProcessor: EncounterFormProcessor,
+  });
+  const getSubForms = () => Object.values(subForms.current);
+  const getRootForm = () => rootForm.current;
+
+  useEffect(() => {
+    if (isSubmitting) {
+      // TODO: find a dynamic way of managing the forms processing order
+      const forms = [rootForm.current, ...getSubForms()];
+      // validate all forms
+      const valid = forms.every((formContext) => validateForm(formContext));
+      if (valid) {
+        Promise.all(
+          forms.map((formContext) => formContext.processor.processSubmission(formContext, new AbortController())),
+        )
+          .then((results) => {
+            // TODO: process post submission actions
+            // TODO: handle form submission success
+            formSubmissionProps.setIsSubmitting(false);
+            formSubmissionProps.handleClose();
+          })
+          .catch((error) => {
+            // TODO: handle form submission errors
+            formSubmissionProps.setIsSubmitting(false);
+          });
+      } else {
+        formSubmissionProps.setIsSubmitting(false);
+      }
+    }
+  }, [isSubmitting]);
+
+  return (
+    <FormFactoryProviderContext.Provider
+      value={{
+        patient,
+        sessionMode,
+        sessionDate,
+        formJson,
+        formProcessors: formProcessors.current,
+        layoutType,
+        workspaceLayout,
+        visit,
+        location,
+        provider,
+        registerForm,
+        getSubForms,
+        getRootForm,
```

Comment:
Is there a use case for providing these methods to children? The only use I see of them is in the "submitting" hook and it's better if we expose as minimal an API as possible to each child.

Reply:
Nice argument! Updating 🧑‍🍳